[P2P] Better peer insights

MithrilMan opened the pull request:
Better peer insights using event bus and a diagnostic feature.
Work In progress.

MithrilMan commented:
I created this while working on https://dev.azure.com/Stratisplatformuk/StratisBitcoinFullNode/_workitems/edit/3489/

I know many of us had problem understanding peer statuses while a node is running, and having to rely on logs is not always an option (and not even easy to understand).

This is why I've created a new feature called Stratis.DiagnosticFeature that could be an handy way to hook into the node internal details and expose informations using API actions.
The first component I've created is what I called `PeerStatisticsCollector` whose goal is to collect peer events and aggregate them in an easy to read output to inspect current status of the node.
To achieve this I've used the event bus to publish peer events.
Their name are self explainatory (and anyway they are commented):
`PeerConnected`, `PeerConnectionAttempt`, `PeerConnectionAttemptFailed`, `PeerDisconnected`, `PeerMessageReceived`, `PeerMessageSendFailure`, `PeerMessageSent`

Each of these events carry their own information (e.g. `PeerDisconnected `has the disconnected reason and eventually the exception)

This implementation may serves as reference for other diagnostic implementation and as an example about how to use properly the EventBus (aka Signaler) because to not impact on performance (event publishing are actually synchronous calls), subscribers to those events put the events in a queue and a separate thread dequeue them (using AsyncProvider dequeuer)

I've updated `Stratis.StratisD` to include this feature but by default PeerStatisticsCollector is turned off so it wouldn't impact in any way in fullnode performance.

to enable the peer statistic collector you can both use command args (-help show available arguments)

> info: Stratis.Bitcoin.Configuration.NodeSettings[0]
>       **DIAGNOSTIC SETTINGS**
>       -diagpeerstats=<bool>             Start the diagnostic peer statistics collector. Defaults to False.
>       -diagpeerstatsmaxlog=<number>     Maximum number of logged peer events stored during the diagnostic peer statistics collector. Defaults to 10.
>       **END OF DIAGNOSTIC SETTINGS**

By default the max number of logged events per peer is set to 10 (can be changed by the argument `-diagpeerstatsmaxlog=nn`)

In swagger you'll find all diagnostic actions available 
![image](https://user-images.githubusercontent.com/5107375/56423056-ab55cc80-62aa-11e9-8638-628f669a5498.png)

As you can see you can manually start and stop peer statistic collection from swagger

MithrilMan commented:
this is an example of the output (I was testing it with 15 max logged events)

[
	{
		"peerEndPoint": "[::ffff:51.140.255.152]:16178",
		"connected": true,
		"inbound": false,
		"bytesSent": 1146284,
		"bytesReceived": 26332192,
		"receivedMessages": 30429,
		"sentMessages": 126,
		"latestEvents": [
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block"
		]
	},
	{
		"peerEndPoint": "[::ffff:95.165.168.232]:16178",
		"connected": true,
		"inbound": false,
		"bytesSent": 1147049,
		"bytesReceived": 27406939,
		"receivedMessages": 30215,
		"sentMessages": 133,
		"latestEvents": [
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block"
		]
	},
	{
		"peerEndPoint": "[2003:c0:7700:9eb7:95a1:18ed:6c03:6179]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:02:27] Attempting Connection",
			"[19/04/2019 12:02:28] Disconnected. Reason: Unexpected exception while connecting to socket. Exception: System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (10051): Tentativo di operazione del socket verso una rete non raggiungibile [2003:c0:7700:9eb7:95a1:18ed:6c03:6179]:16178\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeerConnection.ConnectAsync(IPEndPoint endPoint, CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeerConnection.cs:line 211\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 415\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 450\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeerFactory.CreateConnectedNetworkPeerAsync(IPEndPoint peerEndPoint, NetworkPeerConnectionParameters parameters, NetworkPeerDisposer networkPeerDisposer) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeerFactory.cs:line 206\r\n   at Stratis.Bitcoin.P2P.PeerConnector.ConnectAsync(PeerAddress peerAddress) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\PeerConnector.cs:line 237"
		]
	},
	{
		"peerEndPoint": "[::ffff:92.150.96.185]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:02:29] Attempting Connection",
			"[19/04/2019 12:02:34] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:95.216.14.110]:16178",
		"connected": true,
		"inbound": false,
		"bytesSent": 1406343,
		"bytesReceived": 51636942,
		"receivedMessages": 33885,
		"sentMessages": 259,
		"latestEvents": [
			"[19/04/2019 12:05:25] Message Received: block",
			"[19/04/2019 12:05:25] Message Received: block",
			"[19/04/2019 12:05:25] Message Received: block",
			"[19/04/2019 12:05:25] Message Received: block",
			"[19/04/2019 12:05:25] Message Received: block",
			"[19/04/2019 12:05:26] Message Received: block",
			"[19/04/2019 12:05:26] Message Sent: getdata",
			"[19/04/2019 12:05:27] Message Received: block",
			"[19/04/2019 12:05:27] Message Received: block",
			"[19/04/2019 12:05:27] Message Received: block",
			"[19/04/2019 12:05:27] Message Received: block",
			"[19/04/2019 12:05:27] Message Received: block",
			"[19/04/2019 12:05:27] Message Received: block",
			"[19/04/2019 12:05:28] Message Sent: getdata",
			"[19/04/2019 12:05:28] Message Received: block"
		]
	},
	{
		"peerEndPoint": "[::ffff:185.183.185.76]:16178",
		"connected": true,
		"inbound": false,
		"bytesSent": 23230,
		"bytesReceived": 3608938,
		"receivedMessages": 28,
		"sentMessages": 24,
		"latestEvents": [
			"[19/04/2019 12:04:43] Message Sent: getheaders",
			"[19/04/2019 12:04:54] Message Received: headers",
			"[19/04/2019 12:04:54] Message Sent: getheaders",
			"[19/04/2019 12:05:02] Message Received: headers",
			"[19/04/2019 12:05:02] Message Sent: getheaders",
			"[19/04/2019 12:05:04] Message Received: headers",
			"[19/04/2019 12:05:04] Message Sent: getheaders",
			"[19/04/2019 12:05:11] Message Received: headers",
			"[19/04/2019 12:05:11] Message Sent: getheaders",
			"[19/04/2019 12:05:14] Message Received: headers",
			"[19/04/2019 12:05:14] Message Sent: getheaders",
			"[19/04/2019 12:05:17] Message Received: headers",
			"[19/04/2019 12:05:17] Message Sent: getheaders",
			"[19/04/2019 12:05:27] Message Received: headers",
			"[19/04/2019 12:05:27] Message Sent: getheaders"
		]
	},
	{
		"peerEndPoint": "[::ffff:86.173.221.77]:16178",
		"connected": true,
		"inbound": false,
		"bytesSent": 24283,
		"bytesReceived": 3772951,
		"receivedMessages": 27,
		"sentMessages": 25,
		"latestEvents": [
			"[19/04/2019 12:04:54] Message Received: headers",
			"[19/04/2019 12:04:54] Message Sent: getheaders",
			"[19/04/2019 12:04:56] Message Received: inv",
			"[19/04/2019 12:04:59] Message Received: headers",
			"[19/04/2019 12:04:59] Message Sent: getheaders",
			"[19/04/2019 12:05:04] Message Received: headers",
			"[19/04/2019 12:05:04] Message Sent: getheaders",
			"[19/04/2019 12:05:10] Message Received: headers",
			"[19/04/2019 12:05:10] Message Sent: getheaders",
			"[19/04/2019 12:05:15] Message Received: headers",
			"[19/04/2019 12:05:15] Message Sent: getheaders",
			"[19/04/2019 12:05:20] Message Received: headers",
			"[19/04/2019 12:05:20] Message Sent: getheaders",
			"[19/04/2019 12:05:26] Message Received: headers",
			"[19/04/2019 12:05:26] Message Sent: getheaders"
		]
	},
	{
		"peerEndPoint": "[::ffff:93.40.7.28]:16178",
		"connected": true,
		"inbound": false,
		"bytesSent": 1442264,
		"bytesReceived": 56372317,
		"receivedMessages": 33799,
		"sentMessages": 298,
		"latestEvents": [
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block",
			"[19/04/2019 12:05:28] Message Received: block"
		]
	},
	{
		"peerEndPoint": "[::ffff:185.244.214.216]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:02:36] Attempting Connection",
			"[19/04/2019 12:02:37] Disconnected. Reason: Unexpected exception while connecting to socket. Exception: System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (10061): Impossibile stabilire la connessione. Rifiuto persistente del computer di destinazione [::ffff:185.244.214.216]:16178\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeerConnection.ConnectAsync(IPEndPoint endPoint, CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeerConnection.cs:line 211\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 415\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 450"
		]
	},
	{
		"peerEndPoint": "[::ffff:77.172.158.141]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:02:38] Attempting Connection",
			"[19/04/2019 12:02:39] Disconnected. Reason: Unexpected exception while connecting to socket. Exception: System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (10061): Impossibile stabilire la connessione. Rifiuto persistente del computer di destinazione [::ffff:77.172.158.141]:16178\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeerConnection.ConnectAsync(IPEndPoint endPoint, CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeerConnection.cs:line 211\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 415\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 450\r\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()"
		]
	},
	{
		"peerEndPoint": "[::ffff:163.172.47.92]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 135,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 1,
		"latestEvents": [
			"[19/04/2019 12:02:40] Attempting Connection",
			"[19/04/2019 12:02:40] Peer Connected",
			"[19/04/2019 12:02:40] Message Sent: version",
			"[19/04/2019 12:02:40] Disconnected. Reason: Receiving cancelled.. Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:168.181.51.210]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:02:41] Attempting Connection",
			"[19/04/2019 12:02:46] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:185.133.60.10]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:02:47] Attempting Connection",
			"[19/04/2019 12:02:48] Disconnected. Reason: Unexpected exception while connecting to socket. Exception: System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (10061): Impossibile stabilire la connessione. Rifiuto persistente del computer di destinazione [::ffff:185.133.60.10]:16178\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeerConnection.ConnectAsync(IPEndPoint endPoint, CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeerConnection.cs:line 211\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 415\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 450"
		]
	},
	{
		"peerEndPoint": "[::ffff:142.4.216.199]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 135,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 1,
		"latestEvents": [
			"[19/04/2019 12:02:49] Attempting Connection",
			"[19/04/2019 12:02:49] Peer Connected",
			"[19/04/2019 12:02:49] Message Sent: version",
			"[19/04/2019 12:02:49] Disconnected. Reason: Receiving cancelled.. Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:109.111.156.63]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:02:53] Attempting Connection",
			"[19/04/2019 12:02:58] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[2001:0:9d38:6ab8:2c55:ec6:ace4:a32]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:02:59] Attempting Connection",
			"[19/04/2019 12:02:59] Disconnected. Reason: Unexpected exception while connecting to socket. Exception: System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (10051): Tentativo di operazione del socket verso una rete non raggiungibile [2001:0:9d38:6ab8:2c55:ec6:ace4:a32]:16178\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeerConnection.ConnectAsync(IPEndPoint endPoint, CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeerConnection.cs:line 211\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 415"
		]
	},
	{
		"peerEndPoint": "[::ffff:62.178.219.168]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:03:00] Attempting Connection",
			"[19/04/2019 12:03:05] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:104.211.98.51]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 135,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 1,
		"latestEvents": [
			"[19/04/2019 12:03:06] Attempting Connection",
			"[19/04/2019 12:03:06] Peer Connected",
			"[19/04/2019 12:03:06] Message Sent: version",
			"[19/04/2019 12:03:07] Disconnected. Reason: Receiving cancelled.. Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:104.254.93.67]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:03:08] Attempting Connection",
			"[19/04/2019 12:03:13] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:178.252.9.183]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 270,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 2,
		"latestEvents": [
			"[19/04/2019 12:03:14] Attempting Connection",
			"[19/04/2019 12:03:14] Peer Connected",
			"[19/04/2019 12:03:14] Message Sent: version",
			"[19/04/2019 12:03:19] Disconnected. Reason: Connection timeout. Exception: ",
			"[19/04/2019 12:04:26] Attempting Connection",
			"[19/04/2019 12:04:26] Peer Connected",
			"[19/04/2019 12:04:26] Message Sent: version",
			"[19/04/2019 12:04:31] Disconnected. Reason: Connection timeout. Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:178.25.128.72]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:03:20] Attempting Connection",
			"[19/04/2019 12:03:25] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:40.78.80.159]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 270,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 2,
		"latestEvents": [
			"[19/04/2019 12:03:26] Attempting Connection",
			"[19/04/2019 12:03:26] Peer Connected",
			"[19/04/2019 12:03:26] Message Sent: version",
			"[19/04/2019 12:03:26] Disconnected. Reason: Receiving cancelled.. Exception: ",
			"[19/04/2019 12:04:44] Attempting Connection",
			"[19/04/2019 12:04:45] Peer Connected",
			"[19/04/2019 12:04:45] Message Sent: version",
			"[19/04/2019 12:04:45] Disconnected. Reason: Receiving cancelled.. Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:98.251.124.215]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 135,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 1,
		"latestEvents": [
			"[19/04/2019 12:03:27] Attempting Connection",
			"[19/04/2019 12:03:27] Peer Connected",
			"[19/04/2019 12:03:27] Message Sent: version",
			"[19/04/2019 12:03:32] Disconnected. Reason: Connection timeout. Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:91.121.163.64]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 135,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 1,
		"latestEvents": [
			"[19/04/2019 12:03:33] Attempting Connection",
			"[19/04/2019 12:03:33] Peer Connected",
			"[19/04/2019 12:03:33] Message Sent: version",
			"[19/04/2019 12:03:38] Disconnected. Reason: Connection timeout. Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:80.211.206.193]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 135,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 1,
		"latestEvents": [
			"[19/04/2019 12:03:39] Attempting Connection",
			"[19/04/2019 12:03:39] Peer Connected",
			"[19/04/2019 12:03:39] Message Sent: version",
			"[19/04/2019 12:03:39] Disconnected. Reason: Receiving cancelled.. Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:84.161.49.203]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:03:40] Attempting Connection",
			"[19/04/2019 12:03:45] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:185.64.104.55]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 135,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 1,
		"latestEvents": [
			"[19/04/2019 12:03:46] Attempting Connection",
			"[19/04/2019 12:03:46] Peer Connected",
			"[19/04/2019 12:03:46] Message Sent: version",
			"[19/04/2019 12:03:46] Disconnected. Reason: Receiving cancelled.. Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:213.219.155.18]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:03:47] Attempting Connection",
			"[19/04/2019 12:03:52] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:74.138.34.83]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 135,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 1,
		"latestEvents": [
			"[19/04/2019 12:03:53] Attempting Connection",
			"[19/04/2019 12:03:53] Peer Connected",
			"[19/04/2019 12:03:53] Message Sent: version",
			"[19/04/2019 12:03:58] Disconnected. Reason: Connection timeout. Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:84.168.35.30]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:03:59] Attempting Connection",
			"[19/04/2019 12:04:04] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:142.93.35.42]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 135,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 1,
		"latestEvents": [
			"[19/04/2019 12:04:05] Attempting Connection",
			"[19/04/2019 12:04:05] Peer Connected",
			"[19/04/2019 12:04:05] Message Sent: version",
			"[19/04/2019 12:04:05] Disconnected. Reason: Receiving cancelled.. Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:209.239.123.108]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 135,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 1,
		"latestEvents": [
			"[19/04/2019 12:04:06] Attempting Connection",
			"[19/04/2019 12:04:07] Peer Connected",
			"[19/04/2019 12:04:07] Message Sent: version",
			"[19/04/2019 12:04:07] Disconnected. Reason: Receiving cancelled.. Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:91.224.58.128]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:04:08] Attempting Connection",
			"[19/04/2019 12:04:13] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:94.207.174.27]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:04:14] Attempting Connection",
			"[19/04/2019 12:04:19] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:46.216.17.248]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:04:21] Attempting Connection",
			"[19/04/2019 12:04:22] Disconnected. Reason: Unexpected exception while connecting to socket. Exception: System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (10061): Impossibile stabilire la connessione. Rifiuto persistente del computer di destinazione [::ffff:46.216.17.248]:16178\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeerConnection.ConnectAsync(IPEndPoint endPoint, CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeerConnection.cs:line 211\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 415\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 450"
		]
	},
	{
		"peerEndPoint": "[::ffff:91.131.150.153]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:04:23] Attempting Connection",
			"[19/04/2019 12:04:24] Disconnected. Reason: Unexpected exception while connecting to socket. Exception: System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (10061): Impossibile stabilire la connessione. Rifiuto persistente del computer di destinazione [::ffff:91.131.150.153]:16178\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeerConnection.ConnectAsync(IPEndPoint endPoint, CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeerConnection.cs:line 211\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 415\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 450"
		]
	},
	{
		"peerEndPoint": "[::ffff:158.69.118.2]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 135,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 1,
		"latestEvents": [
			"[19/04/2019 12:04:25] Attempting Connection",
			"[19/04/2019 12:04:25] Peer Connected",
			"[19/04/2019 12:04:25] Message Sent: version",
			"[19/04/2019 12:04:25] Disconnected. Reason: Receiving cancelled.. Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:31.20.112.47]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:04:32] Attempting Connection",
			"[19/04/2019 12:04:37] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:188.0.175.107]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:04:38] Attempting Connection",
			"[19/04/2019 12:04:43] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:99.104.98.112]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:04:46] Attempting Connection",
			"[19/04/2019 12:04:51] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:89.238.154.171]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:04:52] Attempting Connection",
			"[19/04/2019 12:04:53] Disconnected. Reason: Unexpected exception while connecting to socket. Exception: System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (10061): Impossibile stabilire la connessione. Rifiuto persistente del computer di destinazione [::ffff:89.238.154.171]:16178\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeerConnection.ConnectAsync(IPEndPoint endPoint, CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeerConnection.cs:line 211\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 415\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 450"
		]
	},
	{
		"peerEndPoint": "[::ffff:76.25.131.151]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 135,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 1,
		"latestEvents": [
			"[19/04/2019 12:04:54] Attempting Connection",
			"[19/04/2019 12:04:54] Peer Connected",
			"[19/04/2019 12:04:54] Message Sent: version",
			"[19/04/2019 12:04:54] Disconnected. Reason: Receiving cancelled.. Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:149.56.31.230]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 135,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 1,
		"latestEvents": [
			"[19/04/2019 12:04:55] Attempting Connection",
			"[19/04/2019 12:04:55] Peer Connected",
			"[19/04/2019 12:04:55] Message Sent: version",
			"[19/04/2019 12:04:56] Disconnected. Reason: Receiving cancelled.. Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:112.170.30.251]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 135,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 1,
		"latestEvents": [
			"[19/04/2019 12:04:57] Attempting Connection",
			"[19/04/2019 12:04:57] Peer Connected",
			"[19/04/2019 12:04:57] Message Sent: version",
			"[19/04/2019 12:05:02] Disconnected. Reason: Connection timeout. Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:88.156.136.91]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:05:03] Attempting Connection",
			"[19/04/2019 12:05:08] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:146.0.189.17]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:05:09] Attempting Connection",
			"[19/04/2019 12:05:14] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:93.45.228.58]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 0,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 0,
		"latestEvents": [
			"[19/04/2019 12:05:15] Attempting Connection",
			"[19/04/2019 12:05:20] Disconnected. Reason: . Exception: "
		]
	},
	{
		"peerEndPoint": "[::ffff:82.6.182.106]:16178",
		"connected": false,
		"inbound": false,
		"bytesSent": 135,
		"bytesReceived": 0,
		"receivedMessages": 0,
		"sentMessages": 1,
		"latestEvents": [
			"[19/04/2019 12:05:24] Attempting Connection",
			"[19/04/2019 12:05:24] Peer Connected",
			"[19/04/2019 12:05:24] Message Sent: version"
		]
	}
]

zeptin2 commented on src/Stratis.Bitcoin/EventBus/CoreEvents/Peer/PeerMessageSendFailure.cs:
```diff
@@ -0,0 +1,25 @@
+namespace Stratis.Bitcoin.EventBus.CoreEvents
+{
+    /// <summary>
+    /// A peer message has been sent successfully.
```
Comment seems incorrect

zeptin2 commented on src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs:
```diff
@@ -157,16 +162,19 @@ private async Task AcceptClientsAsync()
+                    (bool successful, string reason) connectionAttempt = this.AllowClientConnection(tcpClient);
+                    if (!connectionAttempt.successful)
                     {
+                        this.signals.Publish(new PeerConnectionAttemptFailed(true, (IPEndPoint)tcpClient.Client.RemoteEndPoint, connectionAttempt.reason));
```
Is this perhaps also needed inside the outer `catch` block catching `Exception`? I suppose there isn't much more that can fail after this point in the code.

MithrilMan replied:
mmm I'm not sure, the only way to go in that catch is if this.AllowClientConnection throws or L163, and anyway in that case the NetworkPeerServer AcceptClientsAsync will exit, meaning the server would stop listening for incoming peers...  I don't know why we have chosen to not catch inside the loop and maybe in case of that exception (that 
shouldn't happen) just wait few second before continuing to listen for connection

Who knows, maybe that error sometimes throw and we stop without knowing, probably we should write that in bold... or better use AsyncProvider, could be worthy to open an issue maybe

MithrilMan commented:
I updated some of the code and I've refactored NetworkPeerConnection code, self contained in this commit: https://github.com/stratisproject/StratisBitcoinFullNode/pull/3534/commits/9c1135cd1f7bd57725cd63994fb6444853f55cc0

Probably SocketException and Exception could be merged, but I wanted to highlight Socket Exceptions because they are the ones that happens when the port is closed ( code 10061 ), etc...

zeptin2 commented on src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs:
```diff
@@ -173,40 +176,32 @@ public async Task ConnectAsync(IPEndPoint endPoint, CancellationToken cancellati
-            catch (Exception e)
+            catch (SocketException ex)
             {
-                if (e is AggregateException) e = e.InnerException;
```
I see this line got removed, is that intentional? Unless there is no possibility of an `AggregateException` getting thrown I think it is useful to know the inner exception.

MithrilMan replied:
There is no chance because when you use await, the exception isn't kept in Task but thrown directly
Now I await the `this.tcpClient.ConnectAsync` so the exception handling is easier.

zeptin2 commented on src/Stratis.Features.Diagnostic/PeerDiagnostic/PeerStatistics.cs:
```diff
@@ -0,0 +1,49 @@
+{
+    /// <summary>
+    /// Holds peer statistics.
+    /// <see cref="LatestEvents"/> is a limited size string rapresentation of latest peer events, it's maximum size is specified in the class constructor maxLoggedEvents
```
`representation`, `its`

zeptin2 commented on src/Stratis.Features.Diagnostic/PeerDiagnostic/PeerStatisticsCollector.cs:
```diff
@@ -0,0 +1,169 @@
+                    break;
+                case Event.PeerConnectionAttemptFailed @event:
+                    statistics.Inbound = @event.Inbound;
+                    statistics.LogEvent($"Connection attemp FAILED. Reason: {@event.Reason}.");
```
`attempt`

zeptin2 commented:
I have tried the new API calls myself and it's looking good. Perhaps someone else should take a look too and then we can approve and merge.

mikedennis commented:
Very nice... this is something I can definitely make use of in troubleshooting P2P messaging.

fassadlr commented on src/Stratis.Features.Diagnostic/Controllers/DiagnosticController.cs:
```diff
@@ -0,0 +1,157 @@
+        public DiagnosticController(PeerStatisticsCollector peerDiagnosticCollector, IConnectionManager connectionManager, IConsensusManager consensusManager)
+            : base(connectionManager: connectionManager, consensusManager: consensusManager)
+        {
+            this.peerStatisticsCollector = peerDiagnosticCollector;
```
`peerDiagnosticCollector` should be `peerStatisticsCollector ` ;)

fassadlr commented on src/Stratis.Features.Diagnostic/Controllers/DiagnosticController.cs:
```diff
@@ -0,0 +1,157 @@
+namespace Stratis.Features.Diagnostic.Controllers
+{
+    /// <summary>
+    /// Controller providing operations on a wallet.
```
Summary is wrong.

fassadlr commented on src/Stratis.Features.Diagnostic/Controllers/DiagnosticController.cs:
```diff
@@ -0,0 +1,157 @@
+                {
+                    peersByPeerId = peersByPeerId.Select(DumpPeer),
+                    connectedPeers = connectedPeers.Select(DumpPeer),
+                    connectedPeersNotInPeersByPeerId = connectedPeers.Except(peersByPeerId).Select(DumpPeer)
```
Is it possible to have connected peers in the connection manager that is not in ConsensusManager's peer collection?

MithrilMan replied:
I remember I created this GetConnectedPeersInfo a couple of month ago because this was a condition that was happening sometimes, I don't know if actually it can still happen. Basically I had inconsistent console output where I was reporting to be collected multiple times with the same peer, or not connected to a peer in console but actually I was sending data to him anyway.
Being this a diagnostic feature this method may be useful when we have awkward thing happening, but as I said I'm not sure if this may still happen

fassadlr commented on src/Stratis.Features.Diagnostic/DiagnosticFeature.cs:
```diff
@@ -0,0 +1,75 @@
+            this.peerStatisticsCollector.Dispose();
+        }
+    }
+
```
Extra line

MithrilMan replied:
I had not installed yet all plugins I had on vs 2017, I missed some of this formatting stuffs :/

fassadlr commented on src/Stratis.Features.Diagnostic/DiagnosticSettings.cs:
```diff
@@ -0,0 +1,70 @@
+        {
+            Guard.NotNull(nodeSettings, nameof(nodeSettings));
+
+            this.logger = nodeSettings.LoggerFactory.CreateLogger(typeof(DiagnosticSettings).FullName);
```
`this.logger` should be set to `this.GetType().FullName` ? Then its inline with how we do it in other places.

MithrilMan replied:
yes agree, I copied this setting class from ApiSettings because it was the last setting created, so even that class should be changed

fassadlr commented on src/Stratis.Features.Diagnostic/PeerDiagnostic/PeerStatisticsCollector.cs:
```diff
@@ -0,0 +1,169 @@
+
+                this.peersEventsQueue = this.asyncProvider.CreateAndRunAsyncDelegateDequeuer<Event.PeerEventBase>(nameof(this.peersEventsQueue), UpdatePeerStatistics);
+
+                this.eventSubscriptions.Add(this.signals.Subscribe<Event.PeerConnected>(this.EnqueuePeerEvent));
```
Is there maybe a way that we can just subscribe to all the `PeerEventBase` classes? Or do you want to be explicit.

MithrilMan replied:
We want to be explicit using event bus, will prevent to subscribe to unneeded events and at the same time expresses what we handle

fassadlr commented on src/Stratis.Features.Diagnostic/PeerDiagnostic/PeerStatisticsCollector.cs:
```diff
@@ -0,0 +1,169 @@
+        {
+            return this.peersStatistics.Values.ToList();
+        }
+
```
Extra line

fassadlr commented on src/Stratis.Features.Diagnostic/Stratis.Features.Diagnostic.csproj:
```diff
@@ -0,0 +1,13 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
Perhaps target this to .net core as we will be moving over to it soon?

MithrilMan replied:
yes

MithrilMan replied:
I tried with dotnetcore 2.2 but gave me errors complaining about StratisD not compatible, running 2.1
I'll move this project to netcore2.2 when netcore port will be merged

fassadlr replied:
@MithrilMan We need to still change this to Core 2.1

fassadlr commented on src/Stratis.Features.Diagnostic/Utils/ReflectionExtensionsion.cs:
```diff
@@ -0,0 +1,55 @@
+
+namespace Stratis.Features.Diagnostic.Utils
+{
+    internal static class ReflectionExtensionsion
```
Typo, it should be `ReflectionExtension`

dangershony commented on src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs:
```diff
@@ -173,40 +176,32 @@ public async Task ConnectAsync(IPEndPoint endPoint, CancellationToken cancellati
-                if (error != null)
-                    throw error;
+                this.asyncProvider.Signals.Publish(new PeerConnectionAttempt(false, endPoint));
+                await this.tcpClient.ConnectAsync(endPoint.Address, endPoint.Port).WithCancellationAsync(cancellation).ConfigureAwait(false);
```
`WithCancellationAsync` replace `.Wait(cancellation)`

MithrilMan replied:
pardon?
the core of this change is to not block (always discouraged) and WithCancellationAsync extension solve exactly this problem (now you can await with a timeout)

In the extension comment I left a link to a microsoft blog post that explain this.

dangershony commented on src/Stratis.Bitcoin/Utilities/Extensions/TaskExtension.cs:
```diff
@@ -0,0 +1,49 @@
+        {
+            var tcs = new TaskCompletionSource<bool>();
+
+            using (cancellationToken.Register(s => ((TaskCompletionSource<bool>)s).TrySetResult(true), tcs))
```
We have to be careful here this is a classic memory leak possibility

MithrilMan replied:
where is the memory leak?
It's an `using`, it disposes the token registration automatically
afther the if check

dangershony commented:
Interesting PR but its not really urgent right now

MithrilMan commented:
> Interesting PR but its not really urgent right now

@dangershony I created this PR to try to solve https://dev.azure.com/Stratisplatformuk/StratisBitcoinFullNode/_workitems/edit/3489/ that was (is?) a P1 task, not sure if priority changed but in case they didn't, probably this should go in

fassadlr commented on src/Stratis.Bitcoin/AsyncWork/IAsyncProvider.cs:
```diff
@@ -117,6 +118,15 @@ public interface IAsyncProvider
         string GetStatistics(bool faultyOnly);
 
+        /// <summary>
+        /// temporary hack to expose signals to most of the components (every component currently using asyncprovider), while we decide if we should introduce a component
```
@MithrilMan can you remove this comment and create an issue on the board for something a core component interface (Which I am in favor of btw)

MithrilMan replied:
done
https://dev.azure.com/Stratisplatformuk/StratisBitcoinFullNode/_workitems/edit/4127

fassadlr commented on src/Stratis.Bitcoin/EventBus/CoreEvents/Peer/PeerConnected.cs:
```diff
@@ -0,0 +1,19 @@
+    /// <seealso cref="Stratis.Bitcoin.EventBus.EventBase" />
+    public class PeerConnected : PeerEventBase
+    {
+        public bool Inbound { get; }
```
Perhaps move `Inbound` to the base class? It will apply to all peer messages?

fassadlr commented on src/Stratis.Bitcoin/EventBus/CoreEvents/Peer/PeerMessageReceived.cs:
```diff
@@ -0,0 +1,19 @@
+    /// <seealso cref="Stratis.Bitcoin.EventBus.EventBase" />
+    public class PeerMessageReceived : PeerEventBase
+    {
+        public Message Message { get; }
```
The same here, move `Message` to be base. It can be null :)

MithrilMan replied:
I don't agree on this: PeerConnected/PeerConnectionAttempt/PeerConnectionAttemptFailed/PeerDisconnected doesn't have a concept of Message so having that property, even if null, is not a good design because the model communicate that we may have e Message, that's incorrect

fassadlr commented on src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs:
```diff
@@ -394,7 +395,7 @@ private async Task SetStateAsync(NetworkPeerState newState)
                 {
                     this.logger.LogDebug("Communication with the peer has been closed.");
-
+                    this.asyncProvider.Signals.Publish(new PeerDisconnected(this.Inbound, this.PeerEndPoint, this.DisconnectReason?.Reason, this.DisconnectReason?.Exception));
```
I'm wondering whether we should perhaps only do something like this in debug mode? What do you think?

MithrilMan replied:
These events aren't strictly related to debug (except probably the Message sent/received message that contains the payload) so if we really want to bound these to something, probably should be to the -diagpeerstats argument (or better, the equivalent variable that holds its value because diagnostic can be even enabled by swagger)

I thought already about having them wrapped around an `if` condition when I created this feature but these events can be used by future features like SignalR to relay those events to a GUI to reflect changes on connected peers, etc... and so requiring these use cases to enable debug or diagpeerstats (that indeed add some work, even if negligible because it uses an async loop to handle these events) 


The overlay is really negligible (just instantiate a lightweight class and call a method that will do nothing except checking in a dictionary, because if no one need this events, nothing else will happen) and I don't expect we have thousands of these events per seconds (not even tens in this case because disconnections happens very few times)

We may reconsider this if we see performance issues (I don't expect them tbh) and I'll make use of diagpeerstats to wrap at least messages events in that case, we may even use pooling to prevent to instantiate new messages if needed (but will complicate the design and require the consumers to not store these events references once receaved).

I would be more worried about tons of logger.Debug we have instead :)

dangershony replied:
hmm ok this is not too bad, AsyncProvider is becoming more then a background worker though, here the general component will be more useful.
Injecting Signals is Too much work? I am fine if it stays like this for now.

MithrilMan replied:
Yeah too many classes to touch just to refactor, will cause this PR to contains even more changes, I'd change this once we chose to go (or not to go) with the ICorecomponent thing

fassadlr commented on src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs:
```diff
@@ -136,6 +138,7 @@ private async Task ReceiveMessagesAsync()
                 {
                     Message message = await this.ReadAndParseMessageAsync(this.peer.Version, this.CancellationSource.Token).ConfigureAwait(false);
 
+                    this.asyncProvider.Signals.Publish(new PeerMessageReceived(this.peer.PeerEndPoint, message));
```
The same here, perhaps check if the node is in debug and then only publish?

MithrilMan replied:
same as before

fassadlr commented on src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs:
```diff
@@ -212,18 +220,18 @@ private NetworkPeerConnectionParameters CreateNetworkPeerConnectionParameters()
             {
                 this.logger.LogTrace("(-)[MAX_CONNECTION_THRESHOLD_REACHED]:false");
-                return false;
+                return (false, "Max Connection Threshold Reached.");
```
Lets make these messages more structured. Change the string to: `Inbound Refused: Max Connection Threshold Reached.`

MithrilMan replied:
done

fassadlr commented on src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs:
```diff
@@ -212,18 +220,18 @@ private NetworkPeerConnectionParameters CreateNetworkPeerConnectionParameters()
             {
                 this.logger.LogTrace("(-)[IBD_COMPLETE_ALLOW_CONNECTION]:true");
-                return true;
+                return (true, "IBD Complete, allow connection.");
```
Change the string to: `Inbound Accepted: IBD Complete.`

fassadlr commented on src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs:
```diff
@@ -233,12 +241,12 @@ private bool AllowClientConnection(TcpClient tcpClient)
             {
                 this.logger.LogTrace("(-)[ENDPOINT_WHITELISTED_ALLOW_CONNECTION]:true");
-                return true;
+                return (true, "Endpoint Whitelisted, Allow Connection even if in IBD.");
```
Change the string to: `Inbound Accepted: Endpoint Whitelisted.`

MithrilMan replied:
the concept here is that the endpoint connected even if it was in IBD, because the other peer is whitelisted, the message isn't correct, I changed it to
Inbound Accepted: Whitelisted endpoint connected during IBD.

MithrilMan replied:
nvm I kept your suggestion, it's fine

fassadlr commented on src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs:
```diff
@@ -233,12 +241,12 @@ private bool AllowClientConnection(TcpClient tcpClient)
             this.logger.LogDebug("Node '{0}' is not white listed during initial block download.", clientLocalEndPoint);
 
-            return false;
+            return (false, "Node in IBD, Endpoint not whitelisted, connection refused.");
```
Change the string to: `Inbound Refused: Endpoint not whitelisted during IBD.`

MithrilMan replied:
same as before

fassadlr commented on src/Stratis.Features.Diagnostic/Controllers/DiagnosticController.cs:
```diff
@@ -0,0 +1,157 @@
+        /// <param name="connectedOnly">if set to <c>true</c> returns statistics related to connected peers only.</param>
+        /// <returns>List of peer statistics</returns>
+        [HttpGet]
+        public ActionResult<List<PeerStatisticsModel>> GetPeersStatistics(bool connectedOnly)
```
```suggestion
        public ActionResult<List<PeerStatisticsModel>> GetPeerStatistics(bool connectedOnly)
```

fassadlr commented on src/Stratis.Features.Diagnostic/Controllers/DiagnosticController.cs:
```diff
@@ -0,0 +1,157 @@
+        /// </summary>
+        /// <returns>Operation result.</returns>
+        [HttpGet]
+        public IActionResult StartCollectingPeersStatistics()
```
```suggestion
        public IActionResult StartCollectingPeerStatistics()
```

fassadlr commented on src/Stratis.Features.Diagnostic/Controllers/DiagnosticController.cs:
```diff
@@ -0,0 +1,157 @@
+        /// </summary>
+        /// <returns>Operation result.</returns>
+        [HttpGet]
+        public IActionResult StopCollectingPeersStatistics()
```
```suggestion
        public IActionResult StopCollectingPeerStatistics()
```

fassadlr commented on src/Stratis.Features.Diagnostic/Controllers/Models/PeerStatisticsModel.cs:
```diff
@@ -0,0 +1,42 @@
+{
+    public class PeerStatisticsModel
+    {
+        public string PeerEndPoint { get; set; }
```
@MithrilMan I'm just being pedantic but can you order these fields alphabetically?

fassadlr replied:
Also they should all be `readonly`?

MithrilMan replied:
it's equivalent, they have the getter only

fassadlr commented on src/Stratis.Features.Diagnostic/DiagnosticSettings.cs:
```diff
@@ -0,0 +1,70 @@
+        public const bool DefaultPeersStatisticsCollectorEnabled = false;
+        
+        /// <summary>The default value for maximum peers logged events.</summary>
+        public const int DefaultMaxPeerLoggedEvents = 10;
```
Won't this fill up quickly? Maybe make it a 100?

MithrilMan replied:
my concern was just memory consumption in case of many peers. in my tests it has been enough to diagnostic the problems when I made use of this feature and there is an argument to increase it at will

fassadlr commented on src/Stratis.Features.Diagnostic/PeerDiagnostic/PeerStatistics.cs:
```diff
@@ -0,0 +1,49 @@
+    /// </summary>
+    public class PeerStatistics
+    {
+        public IPEndPoint PeerEndPoint { get; set; }
```
Also, order these properties alphabetically.

fassadlr commented on src/Stratis.Features.Diagnostic/PeerDiagnostic/PeerStatisticsCollector.cs:
```diff
@@ -0,0 +1,168 @@
+                }
+            }
+
+            return statistics;
```
Can we order these statistics by date time?

fassadlr commented on src/Stratis.Features.Diagnostic/Utils/UnityConversionExtension.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace Stratis.Features.Diagnostic.Utils
+{
+    internal static class UnityConversionExtension
```
What is this used for?

MithrilMan replied:
uhm... good point, I'm pretty sure I was using it to format nicely the swagger statistic output but then I chose to display them in raw bytes, so I can remove it

dangershony commented:
@fassadlr I want to review this properly as well please hold pushing till I can review.

MithrilMan commented:
For the record, the problem wasn't related to this PR

fassadlr commented:
@MithrilMan Yes I confirm that the syncing issues arent related to this PR.

Thanks @dangershony  👍

fassadlr commented:
@dangershony @MithrilMan I have managed to sync TestNet to tip and staking works as per normal 👍 Busy syncing MainNet.

dangershony commented:
@fassadlr I will do this today.

dangershony commented on src/Stratis.Bitcoin/AsyncWork/IAsyncProvider.cs:
```diff
@@ -117,6 +118,15 @@ public interface IAsyncProvider
+        /// In favor of ICoreComponents, everytime we need a new core service around, we spend lot of time refactoring every test and many legacy component.
+        /// Having a single entry point for COMMON SERVICES allows us to speed up changes.
+        /// </summary>
+        ISignals Signals { get; }
```
Its becoming an antipattern... but I am not totally against ICoreComponents

MithrilMan replied:
uhm I was sure I had removed that comment, anyway we have an issue about implementing the ICoreComponent (it just doesn't have to be abused)

dangershony commented on src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs:
```diff
@@ -176,40 +176,32 @@ public async Task ConnectAsync(IPEndPoint endPoint, CancellationToken cancellati
+            }
+            catch (Exception ex)
             {
-                if (e is AggregateException) e = e.InnerException;
```
I am not sure .Message is enough in an aggregate exception.

Also you are changing the behaviour here, before it would throw the inner exception (just an FYI).

MithrilMan replied:
forget to reply to this one.
The behavior is the same, just with other code:
`await this.tcpClient.ConnectAsync(endPoint.Address, endPoint.Port).WithCancellationAsync(cancellation).ConfigureAwait(false);` doesn't need to handle an aggregate exception because you need it only when you deal with plain tasks that doesn't throw exceptions, while using the async keyword exception can be catch directly

dangershony replied:
I am not sure I follow, in anycase you aggregate exception can be thrown not only with TPL.
It might not be an issue though.

dangershony commented on src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs:
```diff
@@ -176,40 +176,32 @@ public async Task ConnectAsync(IPEndPoint endPoint, CancellationToken cancellati
             {
-                if (e is AggregateException) e = e.InnerException;
-                this.logger.LogDebug("Error connecting to '{0}', exception message: {1}", endPoint, e.Message);
+                this.asyncProvider.Signals.Publish(new PeerConnectionAttemptFailed(false, endPoint, ex.Message));
```
Why not pass the ex in this case of PeerConnectionAttemptFailed?

MithrilMan replied:
this is raised during connection and from my tests it doesn't give any information better than the message itself (the message could be very long, look for "socket exception" in the output example I've posted).

Consider that these messages are stored in memory when you enable PeerDiagnostic, so I wanted to preserve memory storing only useful data.

If we see that having the full exception object with its stack trace I'm not against passing the exception itself , anyway note that the message contains already some stack information, see the example I'm talking about:

> "[19/04/2019 12:02:37] Disconnected. Reason: Unexpected exception while connecting to socket. Exception: System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (10061): Impossibile stabilire la connessione. Rifiuto persistente del computer di destinazione [::ffff:185.244.214.216]:16178\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeerConnection.ConnectAsync(IPEndPoint endPoint, CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeerConnection.cs:line 211\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 415\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 450"

dangershony replied:
Well if you enabled diagnostics you want to have as much info as possible memory is not a factor in such cases.

dangershony commented on src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs:
```diff
@@ -267,6 +260,7 @@ public async Task SendAsync(Payload payload, CancellationToken cancellation = de
                     byte[] bytes = ms.ToArray();
 
                     await this.SendAsync(bytes, cancellation).ConfigureAwait(false);
+                    this.asyncProvider.Signals.Publish(new PeerMessageSent(this.peer.PeerEndPoint, message, bytes.Length));
```
I realize that this change will create a new instance of the message bug hub even if no one is listening.
Every time a message will be sent (or received in other cases) we create an instance of `PeerMessageSent`

dangershony replied:
I wonder if we should keep regular events for none happy path.

MithrilMan replied:
I'm aware of this and I'm not worried considering the number of events, I consider this a micro optimization that could be done if we ever see performance impact on this aspect (and must be proved by numbers) e.g. using object pooling or other tecniques.

P.S.
can't really compare this with regular events and if you want to mimic its behavior in terms of preserving instance creation, you'd have to check if the eventbus have any subscriber to that event before publishing it and I don't see a nice way without having to call something prior to publish the event and complicate the design... I've an idea tho that I want to experiment with (but as said this will complicate the event bus design a bit)

dangershony replied:
Our node already consumes excessive memory this will only add on the burden and most of the time no one will subscribe to this event.

MithrilMan replied:
consider that the PeerMessageSent class just takes few bytes per messages and got gargabe collected at gen0 if no one consumes it (the message itself if not stored goes out of scope and collected)
if number proves me wrong, will be easy to wrap this call around a property that reflects an argument like -publishpeerevents or something like that

dangershony commented on src/Stratis.Bitcoin/Utilities/Extensions/TaskExtension.cs:
```diff
@@ -0,0 +1,49 @@
+
+namespace Stratis.Bitcoin.Utilities.Extensions
+{
+    public static class TaskExtensions
```
Brilliant

dangershony commented on src/Stratis.CirrusD/Program.cs:
```diff
@@ -56,6 +57,7 @@ private static IFullNode GetSideChainFullNode(NodeSettings nodeSettings)
                 .UseSmartContractWallet()
                 .UseApi()
                 .AddRPC()
+                .UseDiagnosticFeature()
```
Probably not by default, here I am not sure how to proceeded.
If we add diagnostics then it should not be as a feature but a built in part of the node (basically add the components to base) no?

MithrilMan replied:
> Probably not by default, here I am not sure how to proceeded.

note that currently I've implemented in such a way that if you don't run the node with the `-diagpeerstats` argument you do not run any code related to that feature (it just registers the controller that let you know start diagnostic at will and obtain stats) so running it by default its fine.
Only with that argument or running `StartCollectingPeerStatistics` will register to these events and will start an async loop to collect and store them in memory.

> If we add diagnostics then it should not be as a feature but a built in part of the node (basically add the components to base) no?

Then you could say the same thing about wallet, api, rpc or smartcontract.
I even imagine we'll add other diagnostic stuff in this feature in future.
Technically you may want to not include a diagnosticfeature in a node you consider stable, so you are free to not include it in your final assembly (less code, less memory taken by the process, Raspberry Pi will thank you)

Microsoft too decided to create little assemblies for .net framework (e.g. look at how many little assemblies you refer working on .netcore)

Anyway I'm not totally against moving this statistic collector inside base feature, even if I don't see any cons about having its own feature but only pros, feel free to convince me of the opposite :)

dangershony replied:
>I've implemented in such a way that if you don't run the node with the -diagpeerstats argument 
  
Well we will still create instances of events that no one listens to.

>Then you could say the same thing about wallet, api, rpc or smartcontract.  

No I disagree those are products of the fullnode diagnostics is a tool more like configuration or logging.

MithrilMan replied:
> Well we will still create instances of events that no one listens to.
yes, agree, like every logger.Trace/Debug do passing strings and arguments (but called a number of times with higher orders of magnitude)

without considering all the times we capture contexts using lambdas, that generate the instantiations of temporary objects moved between contexts, etc...

We should care of these kind of issues only backed by numbers.
Event Bus works this way on enterprise products, there could be micro optimization at a cost of usabillity or design complexity that actually aren't worth to be taken into account imo.

dangershony commented on src/Stratis.Features.Diagnostic/PeerDiagnostic/PeerStatisticsCollector.cs:
```diff
@@ -0,0 +1,168 @@
+        {
+            this.Enabled = this.diagnosticSettings.PeersStatisticsCollectorEnabled;
+            if (this.Enabled)
+                StartCollecting();
```
this.

MithrilMan replied:
good finding.
If no more important things to fix about this PR, I'll fix this in another one following the merge (builds are failing randomly so having green builds is something to care about :D)

dangershony commented on src/Stratis.Features.Diagnostic/PeerDiagnostic/PeerStatisticsCollector.cs:
```diff
@@ -0,0 +1,168 @@
+        private readonly DiagnosticSettings diagnosticSettings;
+
+        /// <summary>Track current collecting status, when true Peer Collector is collecting statistics.</summary>
+        public bool Enabled { get; private set; }
```
Are you sure this will work? controllers should be statless

dangershony commented:
Generally I like this PR, my only open question is should diagnostics be baked in to base layer or not.

I don't like so much this guy ` .UseDiagnosticFeature() ` diagnostics is disabled by default anyway.

MithrilMan commented:
I replied on each single comment.
As a summary: some concern are micro optimization that should be considered only if numbers says they are a problem (and I don't expect them to be a problem)

About being in base layer or not, see my reply here
https://github.com/stratisproject/StratisBitcoinFullNode/pull/3534#discussion_r305941920

About .UseDiagnosticFeature():
actually this Peer statistic collector is just one of the thing we may want to have in a DiagnosticFeature, so consider this to be just a first step in that direction, so having a UseDiagnosticFeature (with maybe an option to have fine grained control over which diagnostic feature we want to expose to fullnode) makes sense to me.

Note that preventing to publish events related to peer messagging could be easily switched on/off by intrudocing an argument like `-enablePeerEvents` handled in the base layer (so not related strictly to DiagnosticFeature) but as I said I'd take this into account only if we have evidence of performance impact (I could say the same for the thousands of logger.Trace / Debug call we have in code that are called always even if the logger is disabled)

fassadlr commented:
MainNet fully synced.

dangershony commented:
When we create a asp.net selft host server there is nothing similar to `.UseDiagnosticFeature` then why do this here? also you added diagnostics to the main daemon, if anythign its should be not present or we have a diagnostics deamon, or we built it in to the base layer.


`.UseDiagnosticFeature ` is similar to `.UseLogging` or `.UseConfiguration` IMO
